cmd/openfga/openfga_test.go

```diff
@@ -0,0 +1,919 @@
+	require.NoError(t, err, "failed to create openfga docker container")
+
+	stopContainer := func() {
+
+		timeout := 5 * time.Second
+
+		err := dockerClient.ContainerStop(context.Background(), cont.ID, &timeout)
+		if err != nil && !client.IsErrNotFound(err) {
+			t.Fatalf("failed to stop openfga container: %v", err)
+		}
+	}
+
+	err = dockerClient.ContainerStart(context.Background(), cont.ID, types.ContainerStartOptions{})
+	require.NoError(t, err)
+
+	t.Cleanup(stopContainer)
+
+	// spin up a goroutine to survive any test panics or terminations to expire/stop the running container
+	go func() {
+		time.Sleep(2 * time.Minute)
+
+		// swallow the error because by this point we've terminated
+		_ = dockerClient.ContainerStop(context.Background(), cont.ID, nil)
+	}()
+
+	containerJSON, err := dockerClient.ContainerInspect(context.Background(), cont.ID)
+	require.NoError(t, err)
+
+	ports := containerJSON.NetworkSettings.Ports
+
+	m, ok := ports["8080/tcp"]
+	if !ok || len(m) == 0 {
+		t.Fatalf("failed to get HTTP host port mapping from openfga container")
+	}
+	httpPort := m[0].HostPort
+
+	m, ok = ports["8081/tcp"]
+	if !ok || len(m) == 0 {
+		t.Fatalf("failed to get grpc host port mapping from openfga container")
+	}
+	grpcPort := m[0].HostPort
+
+	timeoutCtx, cancel := context.WithTimeout(context.Background(), 3*time.Second)
+	defer cancel()
+
+	creds := insecure.NewCredentials()
+	// if transportCredentials != nil {
+	// 	creds = transportCredentials
+	// }
+
+	dialOpts := []grpc.DialOption{
+		grpc.WithBlock(),
+		grpc.WithTransportCredentials(creds),
+		grpc.WithConnectParams(grpc.ConnectParams{Backoff: grpcbackoff.DefaultConfig}),
+	}
+
+	conn, err := grpc.DialContext(
+		timeoutCtx,
+		fmt.Sprintf("localhost:%s", grpcPort),
+		dialOpts...,
+	)
+	require.NoError(t, err)
+	defer conn.Close()
+
+	client := healthv1pb.NewHealthClient(conn)
+
+	backoffPolicy := backoff.NewExponentialBackOff()
+	backoffPolicy.MaxElapsedTime = 30 * time.Second
+
+	err = backoff.Retry(func() error {
+		timeoutCtx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
+		defer cancel()
+
+		resp, err := client.Check(timeoutCtx, &healthv1pb.HealthCheckRequest{
+			Service: openfgapb.OpenFGAService_ServiceDesc.ServiceName,
+		})
+		if err != nil {
+			return err
+		}
+
+		if resp.GetStatus() != healthv1pb.HealthCheckResponse_SERVING {
+			return fmt.Errorf("not serving")
+		}
+
+		return nil
+	},
+		backoffPolicy,
+	)
+	require.NoError(t, err)
+
+	return &serverHandle{
+		grpcPort: grpcPort,
+		httpPort: httpPort,
+		cleanup:  stopContainer,
+	}, nil
+}
+
+func TestFunctionalGRPC(t *testing.T) {
+
+	require := require.New(t)
+
+	// tester can be shared across tests that aren't impacted
+	// by shared state
+	tester, err := newOpenFGATester(t)
+	require.NoError(err)
+	defer tester.Cleanup()
+
+	t.Run("TestCreateStore", func(t *testing.T) { GRPCCreateStoreTest(t, tester) })
```

Comment:
These tests can kind of be broken down into 2 categories: tests for input validation and tests for standard workflows (functional workflows). The former is a great candidate for test tables, while the later becomes a little more challenging for table driven tests because of the preconditions that have to be met in many cases.

What do you think about creating two separate tests for these different categories? Any preference on this structure?